From a061a8e9adcda6c3a6ae39a82f87da13fc9ee207 Mon Sep 17 00:00:00 2001
From: Maxim Storchak <m.storchak@gmail.com>
Date: Tue, 5 Apr 2022 13:23:06 +0300
Subject: [PATCH] Set loglevel as soon as it's known

Fixes #401

Signed-off-by: Maxim Storchak <m.storchak@gmail.com>
---
 main.c | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/main.c b/main.c
index 166121f2..f382069e 100644
--- a/main.c
+++ b/main.c
@@ -2155,6 +2155,8 @@ int main(int argc, char **argv)
 	if (vpninfo->dump_http_traffic && verbose < PRG_DEBUG)
 		verbose = PRG_DEBUG;
 
+	openconnect_set_loglevel(vpninfo, verbose);
+
 	if (autoproxy) {
 #ifdef LIBPROXY_HDR
 		vpninfo->proxy_factory = px_proxy_factory_new();
@@ -2291,7 +2293,6 @@ int main(int argc, char **argv)
 	}
 
 
-	openconnect_set_loglevel(vpninfo, verbose);
 	openconnect_set_setup_tun_handler(vpninfo, fully_up_cb);
 	openconnect_set_stats_handler(vpninfo, print_connection_stats);
 
-- 
2.25.1

