diff -ur openconnect-5.03.orig/library.c openconnect-5.03/library.c
--- openconnect-5.03.orig/library.c	2014-02-03 14:11:19.000000000 +0100
+++ openconnect-5.03/library.c	2014-08-04 10:07:00.801528724 +0200
@@ -369,6 +369,9 @@
 	if (ret != OATH_OK)
 		return -EIO;
 
+	if (!token_str)
+		return -EINVAL;
+
 	if (strncasecmp(token_str, "base32:", strlen("base32:")) == 0) {
 		ret = oath_base32_decode(token_str + strlen("base32:"),
 					 strlen(token_str) - strlen("base32:"),
